# Numerical Python or Numpy

In [1]:
import numpy as np

In [2]:
import timeit

In [14]:
L = range(1000)
%timeit [i**2 for i in L]

248 µs ± 6.18 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)


# The above took 248 microseconds with list

# numpy is almost 200 times faster and memory efficient used especially for high level calculations 

In [15]:
a = np.arange(1,1000)

%timeit a**2

1.34 µs ± 6.22 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


# Meanwhile the numpy operation took 1.34 microseconds

In [10]:
a

array([1, 2, 3, 4, 5, 6, 7, 8, 9])

In [9]:
a**2 #  double star **means square

array([ 1,  4,  9, 16, 25, 36, 49, 64, 81])

In [16]:
a.ndim

1

In [19]:
a.shape

(999,)

In [27]:
b = np.array([[1,2,3,4],[2,3,4,5,6]])  #2D array

In [21]:
b.shape

(2,)

In [22]:
b.ndim

1

In [25]:
d = np.arange(1,10,2)  #start,stop,skip

In [26]:
d

array([1, 3, 5, 7, 9])

In [34]:
x = np.linspace(0,5,6)  #start,end,number of points

In [35]:
x

array([0., 1., 2., 3., 4., 5.])

In [36]:
z = np.ones(5)

In [37]:
z

array([1., 1., 1., 1., 1.])

In [39]:
np.ones((3,3))

array([[1., 1., 1.],
       [1., 1., 1.],
       [1., 1., 1.]])

In [40]:
np.zeros((3,3))

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [42]:
np.eye(3)     #diagonal matrix

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [43]:
np.eye(3,3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [44]:
c = np.diag([1,2,3,4])

In [45]:
c

array([[1, 0, 0, 0],
       [0, 2, 0, 0],
       [0, 0, 3, 0],
       [0, 0, 0, 4]])

In [46]:
np.random.rand(5)  

array([0.16138999, 0.99551089, 0.6737048 , 0.76620775, 0.98660254])

In [48]:
np.random.randn(5) #gives random samples from a standard normal distribution

array([ 0.76859955,  2.60463866, -0.36235241, -1.10560657,  0.57535508])

In [50]:
np.arange(5,dtype=float)

array([0., 1., 2., 3., 4.])

In [52]:
q = np.array([True,False,True,False])

In [55]:
q.dtype

dtype('bool')

In [58]:
v = np.diag([1,2,3,4])

In [60]:
v

array([[1, 0, 0, 0],
       [0, 2, 0, 0],
       [0, 0, 3, 0],
       [0, 0, 0, 4]])

In [63]:
v[1,1]  #it works like 1st row first column so its 2

2

In [68]:
a = np.random.randint(1,10,10)  #1 to 10 is range and the last one is samples that is 10

In [69]:
a

array([8, 1, 4, 1, 9, 1, 1, 8, 9, 2])

In [75]:
a = np.array([1,2,3,4])

In [77]:
a + 1          #we can add a scalar

array([2, 3, 4, 5])

In [105]:
g = np.array([1,5,6,7])
h = np.array([5,10,11,4])

In [106]:
np.array_equal(g,h)

False

In [98]:
g*h           #array multiplication

array([ 5, 50, 66, 28])

In [101]:
g == h   #array element wise comparison

array([False, False, False, False])

In [104]:
g > h  #again checking if elements of g are greater than h

array([False, False, False,  True])

In [85]:
n = np.diag([1,2,3,4,5])

In [86]:
m= np.diag([5,6,7,8,9])

In [87]:
n.dot(m)           #matrix multiplication is done by dot

array([[ 5,  0,  0,  0,  0],
       [ 0, 12,  0,  0,  0],
       [ 0,  0, 21,  0,  0],
       [ 0,  0,  0, 32,  0],
       [ 0,  0,  0,  0, 45]])

In [108]:
u = np.array([1,2,3,4])

In [110]:
u.sum()

10

In [112]:
#sum by rows and columns 

j = np.array([[1,2,4],
              [4,5,6]])

In [117]:
np.sum(j,axis=1)  #1 is row wise sum here

array([ 7, 15])

In [118]:
np.sum(j,axis=0)    #here its column wise addition

array([ 5,  7, 10])

In [119]:
j.min()

1

In [120]:
j.max()

6

In [121]:
j.argmin()      #argmin gives us indexes on least number

0

In [123]:
j.argmax() #argmax gives us index of the top or max element

5

In [124]:
pwd

'/home/lj'

In [3]:
data = np.loadtxt('populations.txt')

In [5]:
data

array([[ 1900., 30000.,  4000., 48300.],
       [ 1901., 47200.,  6100., 48200.],
       [ 1902., 70200.,  9800., 41500.],
       [ 1903., 77400., 35200., 38200.],
       [ 1904., 36300., 59400., 40600.],
       [ 1905., 20600., 41700., 39800.],
       [ 1906., 18100., 19000., 38600.],
       [ 1907., 21400., 13000., 42300.],
       [ 1908., 22000.,  8300., 44500.],
       [ 1909., 25400.,  9100., 42100.],
       [ 1910., 27100.,  7400., 46000.],
       [ 1911., 40300.,  8000., 46800.],
       [ 1912., 57000., 12300., 43800.],
       [ 1913., 76600., 19500., 40900.],
       [ 1914., 52300., 45700., 39400.],
       [ 1915., 19500., 51100., 39000.],
       [ 1916., 11200., 29700., 36700.],
       [ 1917.,  7600., 15800., 41800.],
       [ 1918., 14600.,  9700., 43300.],
       [ 1919., 16200., 10100., 41300.],
       [ 1920., 24700.,  8600., 47300.]])

In [6]:
#These are the cols in file
year, hares, lynxes, carrots = data.T #transpose T does is it transposes the column as rows and rows as columns

In [8]:
print(year)

[1900. 1901. 1902. 1903. 1904. 1905. 1906. 1907. 1908. 1909. 1910. 1911.
 1912. 1913. 1914. 1915. 1916. 1917. 1918. 1919. 1920.]


In [11]:
population = data[:,1:]       #Here we are slicing and taking all the columns starting from one we are skipping 0th column year

In [17]:
population

array([[30000.,  4000., 48300.],
       [47200.,  6100., 48200.],
       [70200.,  9800., 41500.],
       [77400., 35200., 38200.],
       [36300., 59400., 40600.],
       [20600., 41700., 39800.],
       [18100., 19000., 38600.],
       [21400., 13000., 42300.],
       [22000.,  8300., 44500.],
       [25400.,  9100., 42100.],
       [27100.,  7400., 46000.],
       [40300.,  8000., 46800.],
       [57000., 12300., 43800.],
       [76600., 19500., 40900.],
       [52300., 45700., 39400.],
       [19500., 51100., 39000.],
       [11200., 29700., 36700.],
       [ 7600., 15800., 41800.],
       [14600.,  9700., 43300.],
       [16200., 10100., 41300.],
       [24700.,  8600., 47300.]])

In [18]:
#Mean population
#considering cols other that 1st col i.e year
np.argmax(population,axis=1)         #here axis 1 means we are comparing row wise thus we get the highest population for the same

array([2, 2, 0, 0, 1, 1, 2, 2, 2, 2, 2, 2, 0, 0, 0, 1, 2, 2, 2, 2, 2])

# Broadcasting in Numpy

# Broadcasting
[ [1,2,3], [1,1,1], [3,2,2] ]

+

[ [0,1,2], [0,1,2], [0,1,2] ]

is same as

[ [1,2,3], [1,1,1], [3,2,2] ]

+

[ [0,1,2], ]

it will add 2 more rows to sync up with
first matrix
Other Example:

[ [1], [2], [3], [4] ]

+

[ [0,1,2] ]

is same is

[ [1,2,3,4], [1,2,3,4], [1,2,3,4], [1,2,3,4] ]

+

[ [0,1,2], [0,1,2], [0,1,2], [0,1,2] ]

In [19]:
a = np.tile(np.arange(3),(3,2))   #range till 3 and 3 replicates the same to next row or adjacently like 012 012

print(a);

[[0 1 2 0 1 2]
 [0 1 2 0 1 2]
 [0 1 2 0 1 2]]


In [22]:
a = np.tile(np.arange(3),(3,1))
print("a is ",a)

b = np.arange(3);

print("b is: ",b);               

#here it will add b in every row thus a + b is as follows:

a+b

a is  [[0 1 2]
 [0 1 2]
 [0 1 2]]
b is:  [0 1 2]


array([[0, 2, 4],
       [0, 2, 4],
       [0, 2, 4]])

In [46]:
a = np.array([[1,2,3,4],[4,5,6,7]])

In [47]:
a

array([[1, 2, 3, 4],
       [4, 5, 6, 7]])

In [52]:
t = a.T.ravel()  #transpose with ravel

In [53]:
t

array([1, 4, 2, 5, 3, 6, 4, 7])

In [50]:
t = a.ravel()   #ravel function converts multi dimension array to a single dimension array which is called flattening

In [51]:
t

array([1, 2, 3, 4, 4, 5, 6, 7])

# Reshaping

In [55]:
a.reshape(4,2)

array([[1, 2],
       [3, 4],
       [4, 5],
       [6, 7]])

In [58]:
a.reshape(2,4)  #2 rows 4 columns

array([[1, 2, 3, 4],
       [4, 5, 6, 7]])

In [68]:
q = np.arange(4*3*2).reshape(4,3,2) #to construct a three dimensional matrix 4*3*2 that is 24 elements and reshape into 3d

In [70]:
q  #4 matrices with 3 rows and 2 columns

array([[[ 0,  1],
        [ 2,  3],
        [ 4,  5]],

       [[ 6,  7],
        [ 8,  9],
        [10, 11]],

       [[12, 13],
        [14, 15],
        [16, 17]],

       [[18, 19],
        [20, 21],
        [22, 23]]])

In [73]:
q[0,2,1]   # 0th matrix 2 row and 1 column

5

# sorting along an axis

In [81]:
n = np.array([[0,10,11,12],[2,3,5,1],[5,6,8,7]])

In [89]:
r = np.sort(n,axis=1)

In [90]:
r

array([[ 0, 10, 11, 12],
       [ 1,  2,  3,  5],
       [ 5,  6,  7,  8]])

In [91]:
g = np.array([[0,10,11,12],[2,3,5,1],[5,6,8,7]])

In [92]:
g.sort()

In [93]:
g

array([[ 0, 10, 11, 12],
       [ 1,  2,  3,  5],
       [ 5,  6,  7,  8]])

In [102]:
l = np.array([1,9,8,7])

In [105]:
sorted_address = np.argsort(l)      #we get sorted addresses or indexes

In [106]:
l[sorted_address]         #we put that sorted indexes in our array and get sorted values by putting the sorted indexes

array([1, 7, 8, 9])

In [109]:
z = [1,2,3]

In [110]:
for i in z:
    try:
        if i == 22:
            
            print("lets try")
            
    except:
            print("lets not")

# Try and Except eror handling :

In [117]:
print(x)

NameError: name 'x' is not defined

In [118]:
try:
  print(x)
except NameError:
  print("Variable x is not defined")

Variable x is not defined
